File: src/engines/local/LocalLlamaEngine.ts
import os from "node:os";
import fs from "node:fs";
import {
  getLlama,
  LlamaModel,
  LlamaContext,
  LlamaChatSession,
  ChatMLChatWrapper,
} from "node-llama-cpp";
import { resolveModelPath } from "./modelMap.js";
import { logger } from "../../utils/logging.js";
import { getSession } from "../../session/SessionManager.js";

console.log(`[JARVIS] USING LocalLlamaEngine from ${import.meta.url}`);

export interface GenerationEvent {
  type: "first" | "token" | "done";
  text?: string;
  timestamp: number;
  ms?: number;
}

export interface GenerationParams {
  max_tokens?: number;
  ctx?: number;
  temperature?: number;
  stop?: string[];
}

// Hardware detection and optimization
class HardwareOptimizer {
  public readonly totalRAM: number;
  public readonly availableRAM: number;
  public readonly cpuCores: number;
  public readonly isAppleSilicon: boolean;
  public readonly gpuLayers: number;
  public readonly contextSize: number;
  public readonly batchSize: number;
  public readonly threads: number;
  public readonly modelPath: string;

  constructor() {
    this.totalRAM = Math.floor(os.totalmem() / (1024 * 1024 * 1024));
    this.availableRAM = Math.floor(os.freemem() / (1024 * 1024 * 1024));
    this.cpuCores = os.cpus().length;
    this.isAppleSilicon = process.platform === "darwin" && os.cpus()[0].model.includes("Apple");

    // Detect best model based on RAM
    if (this.totalRAM <= 8) {
      // 8GB or less: Use 3B model
      this.modelPath = "qwen2.5:3b-instruct-q4_K_M";
      this.gpuLayers = this.isAppleSilicon ? 32 : 0; // All layers for 3B
      this.contextSize = 512;
      this.batchSize = 128;
    } else if (this.totalRAM <= 16) {
      // 16GB: Can handle 3B with bigger context or 8B with small context
      this.modelPath = "qwen2.5:3b-instruct-q4_K_M"; 
      this.gpuLayers = this.isAppleSilicon ? 32 : 16;
      this.contextSize = 2048;
      this.batchSize = 256;
    } else {
      // 32GB+: Use 8B model with good settings
      this.modelPath = "llama3.1:8b-instruct-q4_K_M";
      this.gpuLayers = this.isAppleSilicon ? 40 : 20;
      this.contextSize = 4096;
      this.batchSize = 512;
    }

    // CPU threads optimization
    if (this.isAppleSilicon) {
      // Apple Silicon: Use all cores (performance + efficiency)
      this.threads = this.cpuCores;
    } else {
      // x86: Use physical cores only (avoid hyperthreading)
      this.threads = Math.max(1, Math.floor(this.cpuCores / 2));
    }

    logger.info(`🔧 Hardware Optimization:`, {
      ram: `${this.totalRAM}GB (${this.availableRAM}GB free)`,
      cores: this.cpuCores,
      threads: this.threads,
      appleSilicon: this.isAppleSilicon,
      model: this.modelPath,
      gpuLayers: this.gpuLayers,
      context: this.contextSize,
      batch: this.batchSize
    });
  }

  // Dynamic optimization based on current system load
  getDynamicSettings() {
    const currentFreeRAM = Math.floor(os.freemem() / (1024 * 1024 * 1024));
    
    // If system is under memory pressure, reduce context
    if (currentFreeRAM < 2) {
      return {
        contextSize: Math.min(256, this.contextSize),
        batchSize: Math.min(64, this.batchSize),
        threads: Math.max(2, Math.floor(this.threads / 2))
      };
    }
    
    return {
      contextSize: this.contextSize,
      batchSize: this.batchSize,
      threads: this.threads
    };
  }
}

const HARDWARE = new HardwareOptimizer();

export class LocalLlamaEngine {
  private model: LlamaModel | null = null;
  private llama: any = null;

  constructor(private opts?: { modelPath?: string }) {
    logger.info(`Detected ${HARDWARE.isAppleSilicon ? 'Apple Silicon' : 'x86'} with ${HARDWARE.totalRAM}GB RAM`);
  }

  private async ensureModel(modelPath: string) {
    if (this.model) return;

    if (!fs.existsSync(modelPath)) {
      logger.error(`❌ GGUF model not found: ${modelPath}`);
      logger.error("Refusing to proceed without valid model file");
      process.exit(1);
    }

    logger.info(`Loading llama.cpp runtime...`);
    this.llama = await getLlama();

    logger.info(`Loading model from ${modelPath}`);
    
    // Adaptive model loading based on hardware
    const modelConfig: any = {
      modelPath,
      gpuLayers: HARDWARE.gpuLayers,
      useMmap: true,
      useMlock: HARDWARE.totalRAM >= 16, // Lock in RAM only if we have enough
    };

    // Try aggressive settings first, fall back if needed
    try {
      this.model = await this.llama.loadModel(modelConfig);
      logger.info(`Model loaded with ${HARDWARE.gpuLayers} GPU layers`);
    } catch (error) {
      logger.warn("Aggressive settings failed, trying conservative...");
      modelConfig.gpuLayers = Math.floor(HARDWARE.gpuLayers / 2);
      modelConfig.useMlock = false;
      this.model = await this.llama.loadModel(modelConfig);
      logger.info(`Model loaded with ${modelConfig.gpuLayers} GPU layers (conservative)`);
    }
  }

  async *generateStream(
    sessionId: string,
    prompt: string,
    params: GenerationParams
  ): AsyncGenerator<GenerationEvent> {
    const t0 = performance.now();

    // Get dynamic settings based on current system state
    const dynamicSettings = HARDWARE.getDynamicSettings();

    const caps = {
      maxTokens: Math.min(params.max_tokens ?? 128, 512),
      temperature: params.temperature ?? 0.2,
      stop: params.stop ?? ["<|im_end|>", "<|im_start|>", "</s>", "\n\n", "assistant"],
    };
    
    const ctxLen = Math.min(params.ctx ?? dynamicSettings.contextSize, dynamicSettings.contextSize);

    // Prompt budget gate
    if (prompt.length > 4000) {
      if ((process.env.MODE || "dev") === "prod") {
        prompt = "[truncated]" + prompt.slice(-3800);
        logger.warn("Prompt truncated in production mode");
      } else {
        throw new Error(`Prompt too long: ${prompt.length} chars (max 4000 in dev)`);
      }
    }

    // Use hardware-optimized model selection
    const modelPath =
      this.opts?.modelPath ||
      process.env.JARVIS_LOCAL_GGUF ||
      resolveModelPath(HARDWARE.modelPath);

    await this.ensureModel(modelPath);

    // Get or create a persistent context/session for KV reuse
    const { context, chat: session } = await getSession(
      sessionId,
      async () => {
        // Try aggressive context first, reduce if needed
        const sizes = [dynamicSettings.contextSize, 1024, 512, 256];
        
        for (const size of sizes) {
          try {
            const ctx = await this.model!.createContext({
              contextSize: size,
              batchSize: Math.min(dynamicSettings.batchSize, size / 2),
              threads: dynamicSettings.threads,
            });
            logger.info(`Context created: ${size} tokens, ${dynamicSettings.threads} threads`);
            return ctx;
          } catch (error) {
            logger.warn(`Context size ${size} failed, trying smaller...`);
            if (size === sizes[sizes.length - 1]) throw error;
          }
        }
        throw new Error("Failed to create context");
      },
      async (ctx) => {
        const contextSequence = ctx.getSequence();
        const chatSession = new LlamaChatSession({
          contextSequence,
          chatWrapper: new ChatMLChatWrapper(),
        });
        return chatSession;
      }
    );

    logger.debug("[JARVIS] Generation params", {
      ctx: ctxLen,
      batch: dynamicSettings.batchSize,
      threads: dynamicSettings.threads,
      gpuLayers: HARDWARE.gpuLayers,
      model: HARDWARE.modelPath,
    });

    let firstMs: number | null = null;
    let tokensOut = 0;
    let fullResponse = "";

    try {
      const response = await session.prompt(prompt, {
        temperature: caps.temperature,
        maxTokens: caps.maxTokens,
        stopStrings: caps.stop,
      });
      
      if (response) {
        firstMs = performance.now() - t0;
        yield { type: "first", ms: firstMs, timestamp: Date.now() };
        yield { type: "token", text: response, timestamp: Date.now() };
        tokensOut = 1;
        fullResponse = response;
      }
    } catch (e) {
      logger.error("Generation error:", e);
    }

    if (firstMs === null) {
      firstMs = performance.now() - t0;
    }

    const totalMs = performance.now() - t0;

    logger.logLatency({
      engine: "local",
      model_id: HARDWARE.modelPath,
      params: {
        ctx: ctxLen,
        max_tokens: caps.maxTokens,
        temperature: caps.temperature,
        stop: caps.stop,
      },
      prompt_chars: prompt.length,
      session_id: sessionId,
      first_token_ms: firstMs,
      total_ms: totalMs,
      tokens_out: tokensOut,
      route: "local",
    });

    yield { type: "done", timestamp: Date.now() };
  }

  async smokeTest(): Promise<void> {
    logger.info("Running startup smoke test...");

    const testParams: GenerationParams = {
      max_tokens: 16,
      ctx: 512,
      temperature: 0.0,
      stop: [".", "\n", "<|eot_id|>"],
    };

    let firstTokenMs = 0;
    let totalMs = 0;
    let tokensOut = 0;
    let response = "";
    const startTime = performance.now();

    try {
      for await (const event of this.generateStream("smoke-test", "Reply with OK", testParams)) {
        if (event.type === "first") {
          firstTokenMs = event.ms || 0;
        } else if (event.type === "token" && event.text) {
          tokensOut++;
          response += event.text;
        } else if (event.type === "done") {
          totalMs = performance.now() - startTime;
          break;
        }
      }

      logger.info(`Smoke test response: "${response.trim()}"`);

      // Adaptive thresholds based on hardware
      const maxFirstToken = HARDWARE.totalRAM <= 8 ? 8000 : 3000;
      const maxTotal = HARDWARE.totalRAM <= 8 ? 12000 : 5000;

      if (firstTokenMs > maxFirstToken || totalMs > maxTotal || tokensOut === 0) {
        logger.warn(
          `Smoke test slower than ideal: first=${firstTokenMs}ms total=${totalMs}ms tokens=${tokensOut}`
        );
        logger.warn(`Target: first<${maxFirstToken}ms total<${maxTotal}ms`);
        // Don't exit - just warn
      } else {
        logger.info(
          `✅ SMOKE OK first=${firstTokenMs}ms total=${totalMs}ms engine=local tokens=${tokensOut}`
        );
      }
    } catch (error) {
      logger.error("Smoke test failed:", error);
      process.exit(1);
    }
  }

  getHealth() {
    return {
      ok: this.model !== null,
      modelId: HARDWARE.modelPath,
      hardware: {
        ram: `${HARDWARE.totalRAM}GB`,
        cores: HARDWARE.cpuCores,
        threads: HARDWARE.threads,
        appleSilicon: HARDWARE.isAppleSilicon,
      },
      settings: {
        ctx: HARDWARE.contextSize,
        batch: HARDWARE.batchSize,
        gpuLayers: HARDWARE.gpuLayers,
      },
      platform: process.platform,
      metalSupport: HARDWARE.isAppleSilicon,
      engine: "local",
    };
  }

  async cleanup() {
    if (this.model) {
      this.model = null;
      this.llama = null;
      logger.info("Local model cleaned up");
    }
  }
}

File: jarvis/engines/local/local_llama_engine.py
"""Ported from src/engines/local/LocalLlamaEngine.ts (node-llama-cpp -> llama-cpp-python)."""
from __future__ import annotations

import os
import time
from dataclasses import dataclass
from typing import Any, AsyncGenerator, Literal, Optional

from ...utils.logging import logger
from .model_map import get_system_ram_gb, is_apple_silicon, resolve_model_path

GenerationEventType = Literal["first", "token", "done"]


@dataclass
class GenerationEvent:
    type: GenerationEventType
    text: Optional[str] = None
    timestamp: float = 0.0
    ms: Optional[float] = None


@dataclass
class GenerationParams:
    max_tokens: Optional[int] = None
    ctx: Optional[int] = None
    temperature: Optional[float] = None
    stop: Optional[list[str]] = None


class _HardwareOptimizer:
    """Picks model + llama.cpp settings from available RAM, mirroring the
    TS HardwareOptimizer class."""

    def __init__(self) -> None:
        self.total_ram = get_system_ram_gb()
        self.is_apple_silicon = is_apple_silicon()
        self.cpu_threads = os.cpu_count() or 4

        if self.total_ram <= 8:
            self.model_id = "qwen2.5:3b-instruct-q4_K_M"
            self.gpu_layers = 32 if self.is_apple_silicon else 0
            self.context_size = 512
        elif self.total_ram <= 16:
            self.model_id = "qwen2.5:3b-instruct-q4_K_M"
            self.gpu_layers = 32 if self.is_apple_silicon else 16
            self.context_size = 2048
        else:
            self.model_id = "llama3.1:8b-instruct-q4_K_M"
            self.gpu_layers = 40 if self.is_apple_silicon else 20
            self.context_size = 4096

        self.threads = self.cpu_threads if self.is_apple_silicon else max(1, self.cpu_threads // 2)

        logger.info("🔧 Hardware Optimization:", {
            "ram": f"{self.total_ram}GB", "threads": self.threads,
            "appleSilicon": self.is_apple_silicon, "model": self.model_id,
            "gpuLayers": self.gpu_layers, "context": self.context_size,
        })


HARDWARE = _HardwareOptimizer()


class LocalLlamaEngine:
    def __init__(self, model_path: Optional[str] = None) -> None:
        self._model_path_override = model_path
        self._llm: Any = None  # lazily-loaded llama_cpp.Llama instance
        logger.info(f"Detected {'Apple Silicon' if HARDWARE.is_apple_silicon else 'x86'} with {HARDWARE.total_ram}GB RAM")

    def _ensure_model(self, model_path: str) -> None:
        if self._llm is not None:
            return

        if not os.path.exists(model_path):
            logger.error(f"❌ GGUF model not found: {model_path}")
            raise SystemExit("Refusing to proceed without valid model file")

        from llama_cpp import Llama  # imported lazily -- optional heavy dependency

        logger.info(f"Loading model from {model_path}")
        try:
            self._llm = Llama(
                model_path=model_path,
                n_gpu_layers=HARDWARE.gpu_layers,
                n_ctx=HARDWARE.context_size,
                n_threads=HARDWARE.threads,
                use_mmap=True,
                use_mlock=HARDWARE.total_ram >= 16,
            )
            logger.info(f"Model loaded with {HARDWARE.gpu_layers} GPU layers")
        except Exception:  # noqa: BLE001
            logger.warn("Aggressive settings failed, trying conservative...")
            self._llm = Llama(
                model_path=model_path,
                n_gpu_layers=HARDWARE.gpu_layers // 2,
                n_ctx=HARDWARE.context_size,
                n_threads=HARDWARE.threads,
                use_mmap=True,
                use_mlock=False,
            )

    async def generate_stream(
        self, session_id: str, prompt: str, params: GenerationParams
    ) -> AsyncGenerator[GenerationEvent, None]:
        t0 = time.perf_counter()

        max_tokens = min(params.max_tokens or 128, 512)
        temperature = params.temperature if params.temperature is not None else 0.2
        stop = params.stop or ["<|im_end|>", "<|im_start|>", "</s>", "\n\n", "assistant"]
        ctx_len = min(params.ctx or HARDWARE.context_size, HARDWARE.context_size)

        if len(prompt) > 4000:
            if os.environ.get("MODE", "dev") == "prod":
                prompt = "[truncated]" + prompt[-3800:]
                logger.warn("Prompt truncated in production mode")
            else:
                raise ValueError(f"Prompt too long: {len(prompt)} chars (max 4000 in dev)")

        model_path = self._model_path_override or os.environ.get("JARVIS_LOCAL_GGUF") or resolve_model_path(HARDWARE.model_id)
        self._ensure_model(model_path)

        first_ms: Optional[float] = None
        tokens_out = 0
        full_response = ""

        try:
            result = self._llm(
                prompt, max_tokens=max_tokens, temperature=temperature, stop=stop,
            )
            text = result["choices"][0]["text"]
            if text:
                first_ms = (time.perf_counter() - t0) * 1000
                yield GenerationEvent(type="first", ms=first_ms, timestamp=time.time() * 1000)
                yield GenerationEvent(type="token", text=text, timestamp=time.time() * 1000)
                tokens_out = 1
                full_response = text
        except Exception as error:  # noqa: BLE001
            logger.error("Generation error:", error)

        if first_ms is None:
            first_ms = (time.perf_counter() - t0) * 1000

        total_ms = (time.perf_counter() - t0) * 1000

        logger.log_latency({
            "engine": "local", "model_id": HARDWARE.model_id,
            "params": {"ctx": ctx_len, "max_tokens": max_tokens, "temperature": temperature, "stop": stop},
            "prompt_chars": len(prompt), "session_id": session_id,
            "first_token_ms": first_ms, "total_ms": total_ms,
            "tokens_out": tokens_out, "route": "local",
        })

        yield GenerationEvent(type="done", timestamp=time.time() * 1000)

    async def smoke_test(self) -> None:
        logger.info("Running startup smoke test...")
        params = GenerationParams(max_tokens=16, ctx=512, temperature=0.0, stop=[".", "\n"])

        start = time.perf_counter()
        first_token_ms = 0.0
        response = ""
        tokens_out = 0

        async for event in self.generate_stream("smoke-test", "Reply with OK", params):
            if event.type == "first":
                first_token_ms = event.ms or 0
            elif event.type == "token" and event.text:
                tokens_out += 1
                response += event.text
            elif event.type == "done":
                break

        total_ms = (time.perf_counter() - start) * 1000
        max_first_token = 8000 if HARDWARE.total_ram <= 8 else 3000
        max_total = 12000 if HARDWARE.total_ram <= 8 else 5000

        if first_token_ms > max_first_token or total_ms > max_total or tokens_out == 0:
            logger.warn(f"Smoke test slower than ideal: first={first_token_ms}ms total={total_ms}ms tokens={tokens_out}")
        else:
            logger.info(f"✅ SMOKE OK first={first_token_ms}ms total={total_ms}ms engine=local tokens={tokens_out}")

    def get_health(self) -> dict[str, Any]:
        return {
            "ok": self._llm is not None,
            "modelId": HARDWARE.model_id,
            "hardware": {"ram": f"{HARDWARE.total_ram}GB", "threads": HARDWARE.threads, "appleSilicon": HARDWARE.is_apple_silicon},
            "settings": {"ctx": HARDWARE.context_size, "gpuLayers": HARDWARE.gpu_layers},
            "engine": "local",
        }

    async def cleanup(self) -> None:
        self._llm = None
        logger.info("Local model cleaned up")
